web/src/main/java/no/nav/modiapersonoversikt/utils/ConcurrencyUtils.kt

```diff
@@ -0,0 +1,47 @@
+package no.nav.modiapersonoversikt.utils
+
+import no.nav.common.auth.subject.SubjectHandler
+import org.slf4j.MDC
+import org.springframework.web.context.request.RequestAttributes
+import org.springframework.web.context.request.RequestContextHolder
+import java.util.concurrent.CompletableFuture
+
+object ConcurrencyUtils {
+    fun <S, T> inParallel(first: () -> S, second: () -> T): Pair<S, T> {
+        val firstTask = CompletableFuture.supplyAsync(makeThreadSwappable(first))
+        val secondTask = CompletableFuture.supplyAsync(makeThreadSwappable(second))
+
+        CompletableFuture.allOf(firstTask, secondTask).get()
+
+        return Pair(firstTask.get(), secondTask.get())
+    }
+
+    private fun <T> makeThreadSwappable(fn: () -> T): () -> T {
+        val mdc = MDC.getCopyOfContextMap()
+        val subject = SubjectHandler.getSubject()
+        val requestAttributes: RequestAttributes? = RequestContextHolder.getRequestAttributes()
+        return {
+            withRequestAttributes(requestAttributes) {
+                withMDC(mdc) {
+                    SubjectHandler.withSubject(subject.get(), fn)
+                }
+            }
+        }
+    }
+
+    private fun <T> withMDC(mdcContextMap: Map<String?, String?>, fn: () -> T): T {
+        val original = MDC.getCopyOfContextMap()
+        MDC.setContextMap(mdcContextMap)
+        val result = fn()
+        MDC.setContextMap(original)
+        return result
+    }
+
+    private fun <T> withRequestAttributes(requestAttributes: RequestAttributes?, fn: () -> T): T {
+        val original: RequestAttributes? = RequestContextHolder.getRequestAttributes()
+        RequestContextHolder.setRequestAttributes(requestAttributes)
+        val result = fn()
+        RequestContextHolder.setRequestAttributes(original)
+        return result
+    }
+}
```

Comment:
Hva er det som _egentlig_ skjer her? 
MDC er en type logging, er det ikke? Bare prøver å forstå hvorfor vi setter til input, for så å sette til `original` igjen.

Reply:
`MDC`, `SubjectHolder` og `RequestAttributes` er alle lagret ved hjelp av [ThreadLocal](https://docs.oracle.com/javase/7/docs/api/java/lang/ThreadLocal.html), som i bunn og grunn er en fancy måte lage "statiske verdier" for en tråd. Konseptueltsett kan man tenkte på ThreadLocals som `Map<Thread, VALUE>`.

MDC står for "Mapped Diagnostic Context" og brukes for å enrike loggmeldinger. Hos oss bruker vi det blant annet til å lagre unna `callId`, og på den måten kan vi enklere søke etter alle logglinjer med samme `callId`, `requestId` osv.

`RequestAttributes` er spring-web greier som holder styr på ulike attributter relatert til rest-kallet man arbeider med å besvare, f.eks er det mulig å hente ut den underliggende `HttpServletRequest` objektet herifra.

`SubjectHolder` er den som holder styr hvem som er "logget inn".

---

Så kodesnuttene her er for å bevare disse verdiene når man bytter hvilken tråd man kjører på.
Når en http request kommer til modia vil den bli kjørt på en tråd så sant vi ikke gjør noe. Men når vi ønsker å gjøre to ting i parallell så må vi ta ibruk flere tråder. For at koden som kjøres på disse andre trådene fortsatt skal ha tilgang til MDC, Subject etc, så må vi kopiere over disse verdiene. 

Hvis vi titter på `makeThreadSwappable` så er det muligens enklere å forstå:
```kotlin
    fun main() {
      // La oss late som om vi håndterer ett http-kall her, og tråden heter "Thread-0"
      
      // Vi oppretter to "tasks" (lambda funksjoner) som vi ønsker å kjøre i parallell
      // Disse kallene er muligens avhengig av `MDC` o.l, så om vi flyttet de til en annen tråd vil de feile
      val taskA = { slowService.fetchData() }
      val taskB = { slowService.fetchMode() }
      
      // Vi gjør dem derfor "threadswappable"
      val threadIndependentTaskA = makeThreadSwappable(oppgaveA)
      val threadIndependentTaskB = makeThreadSwappable(oppgaveB)
      
      // Vi kan nå kalle de i parallel (på nye tråder) 
      val (resultA, resultB) = inParallel(threadIndependentTaskA, threadIndependentTaskB)
      
    }

    private fun <T> makeThreadSwappable(fn: () -> T): () -> T {
        // Vi er fortsatt på "Thread-0" når denne koden blir kalt, siden det er først i `inParallel` man bytter tråd.
        // Vi kan derfor hente ut verdier fra `ThreadLocal` uten problem, og lagre disse unna i noen lokale variabler.
        val mdc = MDC.getCopyOfContextMap()
        val subject = SubjectHandler.getSubject()
        val requestAttributes: RequestAttributes? = RequestContextHolder.getRequestAttributes()
        
        // Vi returnerer en ny lambda funksjon med samme signatur som vi fikk inn. Fortsatt på "Thread-0"
        return {
           // Ved kjør av denne lambdaen så vil vi ha byttet tråd til "Thread-1". Kall til MDC vil derfor feile om vi ikke gjør noe.
           // Men pga "lexical scoping" har vi fortsatt tilgang til variablene lagret unna like ovenfor.
            withRequestAttributes(requestAttributes) {
                withMDC(mdc) {
                    SubjectHandler.withSubject(subject.get(), fn)
                }
            }
        }
    }

    private fun <T> withRequestAttributes(requestAttributes: RequestAttributes?, fn: () -> T): T {
        // Java utnytter i mange tilfeller ulike ThreadPool, dvs at tråden man er på ikke nødvendig blir avsluttet når man er ferdig med å bruke den.
        // Vi tar derfor ansvar for å rydde opp etter oss. 
        // For å gjøre det henter vi ut nåværende verdier i tilfelle det var noe her ifra før av.
        val original: RequestAttributes? = RequestContextHolder.getRequestAttributes()
        
        // Vi setter så vår nye verdi, hentet mens vi fortsatt var på "Thread-0"
        RequestContextHolder.setRequestAttributes(requestAttributes)
        val result = fn()
        // Og setter tilbake  verdien vi originalt fant her på "Thread-1"
        RequestContextHolder.setRequestAttributes(original)
        return result
    }
```

Reply:
Super forklaring!